routers/api/v1/utils/git.go

```diff
@@ -0,0 +1,59 @@
+// Copyright 2021 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package utils
+
+import (
+	"net/http"
+
+	"code.gitea.io/gitea/modules/context"
+	"code.gitea.io/gitea/modules/git"
+)
+
+// ResolveRefOrSha resolve ref to sha if exist
+func ResolveRefOrSha(ctx *context.APIContext, ref string) string {
+	if len(ref) == 0 {
+		ctx.Error(http.StatusBadRequest, "ref not given", nil)
+		return ""
+	}
+
+	// Search branches and tags
+	for _, refType := range []string{"heads", "tags"} {
+		refSHA, lastMethodName, err := searchRefCommitByType(ctx, refType, ref)
+		if err != nil {
+			ctx.Error(http.StatusInternalServerError, lastMethodName, err)
+			return ""
+		}
+		if refSHA != "" {
+			return refSHA
+		}
+	}
+	return ref
+}
+
+// GetGitRefs return git references based on filter
+func GetGitRefs(ctx *context.APIContext, filter string) ([]*git.Reference, string, error) {
+	gitRepo, err := git.OpenRepository(ctx.Repo.Repository.RepoPath())
```

Comment:
should check if the ctx.Repo.GitRepo is set and if so use that

Reply:
well it original did not work this way but will add this optimisation